### PR TITLE
Merged 'purge' command with 'clean' and deleted 'purge' 

### LINKED ISSUE #2942: Should we merge spack clean and spack purge?
I've talked to some users who mix up `spack clean` and `spack purge`. They basically do the same thing: one deletes the stage directory for a package spec (or multiple package specs) and the other deletes the stage directory for all packages. In light of #2867, should we just merge these into a single command?

### PR DESCRIPTION
**Pull request by alalazo**
fixes #2942

`spack purge` has been merged with `spack clean`. Documentation has been updated accordingly. The `clean` and `purge` behavior are not mutually exclusive, and they log brief information to `tty` while they go.

### CODE REVIEW

File: lib/spack/docs/packaging_guide.rst [hunk 1]
```diff
@@ -3453,6 +3442,12 @@ directory, including cached virtual indices.
 
 To remove all of the above, the command can be called with ``--all``.
 
+Also, cleans up temporary files for a particular package, by deleting the
+expanded/checked out source code *and* any downloaded archive.  If
+``fetch``, ``stage``, or ``install`` are run again after this, Spack's
+build process will start from scratch.
```
Comment (adamjstewart) [inline lib/spack/docs/packaging_guide.rst]: I'm not sure if I would expect `spack clean foo` to delete the downloaded archive. Does `spack clean` delete the downloaded archive for everything? If not, we should keep them in line. I think this would be a good setup:
```
$ spack clean --stage foo  # delete stage for foo
$ spack clean foo  # same as above
$ spack clean --downloads foo  # delete downloaded archive for foo
$ spack clean --stage --downloads foo  # both of the above
```
It would actually be really nice if `spack clean --downloads foo` worked. It makes it a lot easier to test fetching for packages I've already downloaded. I've been manually deleting archives for so long.

There's also a reference above that should be updated.

> When called with ``--stage`` or without arguments this removes all staged files and will be equivalent to running ``spack clean`` for every package you have fetched or staged.

This is no longer _equivalent_ to `spack clean`, it _is_ `spack clean`.

Comment (alalazo) [inline lib/spack/docs/packaging_guide.rst]: The functionality didn't change at all. The command:
```
$ spack clean --stage foo
```
in this branch is equivalent to:
```console
$ spack clean foo
$ spack purge --stage
```
in `develop` (so it deletes the stage for foo, and then it deletes all stages.). Same for other commands. This is reflected in the command help (hopefully):
```
$ spack help clean
usage: spack clean [-h] [-s] [-d] [-m] [-a] ...

remove temporary build files and/or downloaded archives

positional arguments:
  specs             removes the build stages and tarballs for specs

optional arguments:
  -h, --help        show this help message and exit
  -s, --stage       remove all temporary build stages (default)
  -d, --downloads   remove cached downloads
  -m, --misc-cache  remove long-lived caches, like the virtual package index
  -a, --all         equivalent to -sdm
```

----

Good catch for the docs. I'll update the reference.

Comment (adamjstewart) [inline lib/spack/docs/packaging_guide.rst]: I'm not sure if that's true:
```
$ spack clean zlib
$ spack fetch zlib
==> Using cached archive: /Users/Adam/spack/var/spack/cache/zlib/zlib-1.2.11.tar.gz
```
`spack clean foo` doesn't delete the archived tarball. Or are we talking about different tarballs? Maybe you're talking about the one copied to the stage? I was talking about the one in the downloads mirror.

Anyway, I would like to see this behavior:
```
$ spack clean --stage foo  # delete stage for foo
$ spack clean foo  # same as above
$ spack clean --downloads foo  # delete downloaded archive for foo
$ spack clean --stage --downloads foo  # both of the above
```
As long as I can do this, I'm happy. Not sure how much work would be involved in `spack clean --downloads foo` though.

Comment (alalazo) [inline lib/spack/docs/packaging_guide.rst]: Can we keep this extension for another PR? The code here behaves exactly like the code in `develop`, it just merges the two commands. Example:
```console
$ spack clean --stage zlib
==> Cleaning build stage [zlib@1.2.11%gcc@4.8+pic+shared arch=linux-ubuntu14-x86_64 /pgxsxv7]
==> Removing all temporary build stages

$ spack clean --all
==> Removing all temporary build stages
==> Removing cached downloads
==> Removing cached information on repositories
```

File: lib/spack/docs/packaging_guide.rst [hunk 2]
```diff
@@ -3453,6 +3441,13 @@ directory, including cached virtual indices.
 
 To remove all of the above, the command can be called with ``--all``.
 
+When called with positional arguments, cleans up temporary files only
+for a particular package, by deleting the
+expanded/checked out source code *and* any downloaded archive.  If
```
Comment (adamjstewart) [inline lib/spack/docs/packaging_guide.rst]: Just to clarify, does `spack clean foo` delete the stage and the cached download archive? Or just the former?

File: lib/spack/docs/packaging_guide.rst (hunk 1 above)
Comment (adamjstewart) [inline lib/spack/docs/packaging_guide.rst]: I'll open an issue after this is merged.

File: lib/spack/docs/packaging_guide.rst (hunk 2 above)
Comment (alalazo) [inline lib/spack/docs/packaging_guide.rst]: If think what was meant is any downloaded archive *in the stage*. I didn't check, but maybe these documentation precedes the introduction of the cached download archive. Any suggestion for writing this part better is welcome.

Comment (adamjstewart) [inline lib/spack/docs/packaging_guide.rst]: I would just remove the second half of the sentence. We've already explained what `spack clean`/`spack clean --staged` does. Here we are clarifying that `spack clean foo` only applies to `foo`, not to every package.

**Comment by adamjstewart** [comment]
Oh, don't forget to update the Bash completion script!

**Comment by alalazo** [comment]
@tgamblin If you agree with the removal of `spack purge`, I think this is ready to be merged.

**Comment by alalazo** [comment]
@davydden I was wondering if you could do a test drive of this and provide some feedback?

**Comment by alalazo** [comment]
@davydden That's because neither `spack clean` nor `spack purge` were considered basic commands. You should see it doing:
```console
$ spack help -a
```

**Comment by davydden** [comment]
> That's because neither spack clean nor spack purge were considered basic commands

did not know that! Thanks.

**Comment by tgamblin** [comment]
Should we put `spack clean` in the basic help?

**Comment by becker33** [comment]
Not sure we should add it, but I think we could make clearer in the basic help that there are additional commands not listed.

**Comment by becker33** [comment]
@alalazo in #2942 you mentioned wanting to use anonymous specs in `spack clean`. Did you decide these belong in separate PRs?

**Comment by alalazo** [comment]
@becker33 Kind of. I realized that I missed an important detail back then: we don't maintain a db for the build stages as we do for installed packages. 

Without having a db it's difficult to know what is in the stage  and select directories based on the constraints given by the anonymous specs. We can of course maintain a simplified db in the stage, but I am not sure it is worth the effort. Thoughts?

@tgamblin 
> Should we put spack clean in the basic help?

Fine with me. If everybody agrees I can change the group here.

**Comment by becker33** [comment]
I'd agree that it's not worth it, or at least can be a separate issue.

If we wanted to do it we could add things to the database when staged and then mark them installed when installed, but I wouldn't necessarily go for it right now.
